### ISSUE TITLE
header, cpp file and namespace structure / library organisation

### ISSUE THREAD

**Issue by damdamce** [isocpp/CppCoreGuidelines#1259]
This partly overlaps with #497, but imo it's broader, so opening another issue.

I wonder whether it would make sense to make guidelines for structuring the headers and cpp files of libraries. For instance currently
* the stl has toplevel headers (<vector>), but a std:: namespace
* opencv uses <opencv2/core/mat.hpp> and <opencv2/core.hpp>, and a cv:: namespace
* qt uses toplevel <QString> but alternatively also <QtCore/QString>, and the Q prefix (the prefix afaik for historical reasons)

Imo there should be guidelines for:
1. Where to put the header files (top directory (stl) vs. library dir (opencv), that's mainly a packaging and project setup thing, but why not provide guidelines)?
2. Should there be a general namespace for the company, library etc? Should it be long and descriptive or short (2-3 letters)?
3. how to organise functionality
3.1 should namespace names follow the directory structure?
3.2 when should i put a group of classes into a subdirectory?
3.3 when should i use namespaces, separate classes or subclasses?
4. build-systems, specifying that there should be no conflicts when building two .cpp files with the same name but a different path (out of scope?)

examples for 3.:
1. say, i have an abstract base class and many derived classes. Should the derived classes go into a subdirectory? What about the base class? What about naming? e.g. for shapes, should it be "abstract_shape, circle_shape, triangle_shape, etc." or rather "shape::base, shape::circle, shape::triangle" etc with a namespace and directory shape?
2. say i have building blocks, each with certain repetitive classes "shape::object, shape::settings, shape::factory etc.". Should there be a namespace shape as in the example, separate classes "shape, shape_settings, shape_factory etc.", or subclasses of shape "shape, shape::settings, shape::factory etc."?

**Comment by jwakely** [comment]
There's no one-size-fits-all guideline. A single person project with a few files and classes can use any style. A large team with hundreds of files should use whatever works best for the team (possibly structuring the code according to ownership of modules within the team).

This is dealt with in detail in the book _Large Scale C++ Software Design_, see http://www.cs.unc.edu/~stotts/COMP204/lakos/ for an overview of the physical design rules.

**Comment by cubbimew** [comment]
This is essentially our old second [To-do proto-rule](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#S-unclassified) , suggested to be abandoned in https://github.com/isocpp/CppCoreGuidelines/issues/811#issuecomment-299991479

**Comment by damdamce** [comment]
yes, physical design. I hope it'll be covered when modules are there :)

what about part 4. though?

**Comment by hsutter** [comment]
Editors call: We'll postpone this until we have modules broadly available. For now there are too many existing practices, none obviously universally best. We don't want to say anything about build systems.
